### PR TITLE
[macOS] Make FlutterEngine support multiple views

### PR DESCRIPTION
**Pull request by dkwingsmt**
This PR adds support for multiview to `FlutterEngine`. Objective-C user code can add more than one view controllers to`FlutterEngine` as well as removing them.

However, through discussion with goderbauer@ and loic-sharma@, we've decided that publicly `FlutterEngine` will only handle one (unmanaged) view controller, so that it looks exactly the same to the native code by having a `setViewController` that always points to the same view ID. `FlutterEngine` will still support multi-view internally, and a future class will expose this functionality.

## Pre-launch Checklist

- [ ] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [ ] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [ ] I read and followed the [Flutter Style Guide] and the [C++, Objective-C, Java style guides].
- [ ] I listed at least one issue that this PR fixes in the description above.
- [ ] I added new tests to check the change I am making or feature I am adding, or Hixie said the PR is test-exempt. See [testing the engine] for instructions on writing and running engine tests.
- [ ] I updated/added relevant documentation (doc comments with `///`).
- [ ] I signed the [CLA].
- [ ] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[C++, Objective-C, Java style guides]: https://github.com/flutter/engine/blob/main/CONTRIBUTING.md#style
[testing the engine]: https://github.com/flutter/flutter/wiki/Testing-the-engine
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

**Comment by dkwingsmt** [comment]
This PR is ready for review.

File: shell/platform/darwin/macos/framework/Headers/FlutterEngine.h
Comment (loic-sharma) [inline shell/platform/darwin/macos/framework/Headers/FlutterEngine.h:11]:
```suggestion

#include <stdint.h>
```

File: shell/platform/darwin/macos/framework/Headers/FlutterEngine.h
Comment (loic-sharma) [inline shell/platform/darwin/macos/framework/Headers/FlutterEngine.h:84]: Consider something like:

```suggestion
 * The default view is the first view added to the engine, always has ID
 * kFlutterDefaultViewId, and is used by legacy APIs that assume a single view.
```

File: shell/platform/darwin/macos/framework/Headers/FlutterEngine.h
Comment (loic-sharma) [inline shell/platform/darwin/macos/framework/Headers/FlutterEngine.h:95]: Are properties referred as methods in Objective-C?

```suggestion
 * This property is deprecated. Querying view controllers should use
```

File: shell/platform/darwin/macos/framework/Headers/FlutterEngine.h
Comment (loic-sharma) [inline shell/platform/darwin/macos/framework/Headers/FlutterEngine.h:116]:
```suggestion
 * Dissociate the given view controller from this engine.
```

File: shell/platform/darwin/macos/framework/Headers/FlutterViewController.h
Comment (loic-sharma) [inline shell/platform/darwin/macos/framework/Headers/FlutterViewController.h:94]:
```suggestion
 * The initialized view controller will add itself to the engine as part of this process.
```

File: shell/platform/darwin/macos/framework/Source/FlutterEngine.mm
Comment (loic-sharma) [inline shell/platform/darwin/macos/framework/Source/FlutterEngine.mm:723]: What happens if my app creates a view, transitions back to headless mode, then creates another view? Should we guarantee that at least one view has `kFlutterDefaultViewId` if the engine is not in headless mode?

For example, imagine my app is a popup from the menu bar. It starts in headless mode until the user opens the popup. The app returns to headless mode once the user closes the popup. (This is the Google One macOS app scenario)

File: shell/platform/darwin/macos/framework/Source/FlutterEngine.mm
Comment (loic-sharma) [inline shell/platform/darwin/macos/framework/Source/FlutterEngine.mm:93]: Is `withId:` more idiomatic? Same comment for `deregisterViewControllerForId` below

```suggestion
- (void)registerViewController:(FlutterViewController*)controller withId:(uint64_t)viewId;
```

File: shell/platform/darwin/macos/framework/Source/FlutterEngine.mm
Comment (loic-sharma) [reply shell/platform/darwin/macos/framework/Source/FlutterEngine.mm:481]: Same feedback for `viewControllerForId` below

File: shell/platform/darwin/macos/framework/Source/FlutterEngine.mm
Comment (loic-sharma) [inline shell/platform/darwin/macos/framework/Source/FlutterEngine.mm:703]: Should this change also update the embedder API's `update_semantics_callback` callback?

**Comment by loic-sharma** [comment]
@dkwingsmt Could you make sure to include me as a reviewer to any multi-view related engine pull requests? I need to do similar changes in the Windows platform, it's useful for me to track how you implement things on macOS.

**Comment by dkwingsmt** [comment]
Absolutely! Thank you for your careful reviews, they're very helpful.

File: shell/platform/darwin/macos/framework/Headers/FlutterEngine.h
Comment (goderbauer) [inline shell/platform/darwin/macos/framework/Headers/FlutterEngine.h:95]: Consider moving this snippet further up in the doc? I have to read about all the things this can do before I learn that I shouldn't use it.

Comment (goderbauer) [reply shell/platform/darwin/macos/framework/Headers/FlutterEngine.h:95]: Also, there is no official annotation to mark something as deprecated that tools will pick up on?

File: shell/platform/darwin/macos/framework/Headers/FlutterEngine.h
Comment (goderbauer) [reply shell/platform/darwin/macos/framework/Headers/FlutterEngine.h:113]: +1 to this. We should also document that limitation in the doc comment.

File: shell/platform/darwin/macos/framework/Headers/FlutterPluginRegistrarMacOS.h
Comment (goderbauer) [inline shell/platform/darwin/macos/framework/Headers/FlutterPluginRegistrarMacOS.h:44]: Same comment as in the other file about the deprecation notice.

File: shell/platform/darwin/macos/framework/Headers/FlutterEngine.h
Comment (goderbauer) [inline shell/platform/darwin/macos/framework/Headers/FlutterEngine.h:108]: I would consider this an implementation detail and not document this. People shouldn't assume that a view will get a certain ID. They should always read the ID from the view controller.

File: shell/platform/darwin/macos/framework/Source/FlutterEngine.mm
Comment (goderbauer) [reply shell/platform/darwin/macos/framework/Source/FlutterEngine.mm:723]: Overall, I think we should make fewer assumptions about what ID a view will have or get. Ideally, we make no assumptions at all about this. Once consumers of these APIs have migrated to the multi-view APIs they should really check what ID they got instead of relying on some default behavior.

File: shell/platform/darwin/macos/framework/Source/FlutterViewProvider.h
Comment (goderbauer) [inline shell/platform/darwin/macos/framework/Source/FlutterViewProvider.h:23]: Do we have to add this? Ideally, we should no longer make the assumption that there is a default view. If we have this, it bags the question: What is the default view and why is it special?

**Comment by goderbauer** [comment]
Overall it seems a bit strange that we add these public-facing APIs without having them actually be backed by an implementation. See also this comment: https://github.com/flutter/engine/pull/37976#discussion_r1093595076.

File: shell/platform/darwin/macos/framework/Headers/FlutterEngine.h
Comment (cbracken) [inline shell/platform/darwin/macos/framework/Headers/FlutterEngine.h:9]: This is already present on line 11

File: shell/platform/darwin/macos/framework/Headers/FlutterEngine.h
Comment (cbracken) [reply shell/platform/darwin/macos/framework/Headers/FlutterEngine.h:95]: A getter and for mutable properties, a setter method are synthesised, but properties should be referred to as properties in docs.

File: shell/platform/darwin/macos/framework/Headers/FlutterEngine.h
Comment (cbracken) [inline shell/platform/darwin/macos/framework/Headers/FlutterEngine.h:99]:
```suggestion
@property(nonatomic, nullable, weak) FlutterViewController* viewController FLUTTER_DEPRECATED("Use viewControllerForId: instead");
```

File: shell/platform/darwin/macos/framework/Headers/FlutterEngine.h
Comment (cbracken) [reply shell/platform/darwin/macos/framework/Headers/FlutterEngine.h:95]: We have a macro for this. Added a suggestion below.

File: shell/platform/darwin/macos/framework/Headers/FlutterPluginRegistrarMacOS.h
Comment (cbracken) [inline shell/platform/darwin/macos/framework/Headers/FlutterPluginRegistrarMacOS.h:47]:
```suggestion
- (nullable NSView*)view FLUTTER_DEPRECATED("Use viewForId: instead");
```

File: shell/platform/darwin/macos/framework/Source/FlutterCompositorTest.mm
Comment (cbracken) [inline shell/platform/darwin/macos/framework/Source/FlutterCompositorTest.mm:14]: I'm assuming the reason this is a const global rather than constexpr is that it's part of our public headers and therefore we need it to be pure Obj-C and not Obj-C++?

File: shell/platform/darwin/macos/framework/Source/FlutterEngine.mm
Comment (cbracken) [reply shell/platform/darwin/macos/framework/Source/FlutterEngine.mm:93]: `For` is better here. In Obj-C, typically you'll see "withFoo" in constructors and class factory methods (`initWithView:` etc.). For getters, you'll see a lot of forFoo: (e.g. `getFirstRectForRange:`, `getObjectValue:forString:range:error:`).

File: shell/platform/darwin/macos/framework/Source/FlutterCompositorTest.mm
Comment (cbracken) [inline shell/platform/darwin/macos/framework/Source/FlutterCompositorTest.mm:39]: Don't prefix getters with 'get' in Obj-C. `viewForId:...`

File: shell/platform/darwin/macos/framework/Source/FlutterViewProvider.h
Comment (cbracken) [inline shell/platform/darwin/macos/framework/Source/FlutterViewProvider.h:30]: Getters don't take a 'get' prefix in Obj-C. Use `viewForId:...`

File: shell/platform/darwin/macos/framework/Source/FlutterViewProvider.h
Comment (cbracken) [reply shell/platform/darwin/macos/framework/Source/FlutterViewProvider.h:23]: Agreed - I'd avoid this. It feels like adding new API for a legacy concept. Since this isn't public API, it's probably actually advantageous to force the longhand version of looking it up by default ID where needed, just to make the code more explicit and to avoid people accidentally adding more use of single-view assumptions/conveniences.

If, for whatever reason, we somehow absolutely can't live without this method, it should be called `defaultView` not `getDefaultView`. Getter's don't take a leading 'get' in Obj-C.

File: shell/platform/darwin/macos/framework/Source/FlutterViewEngineProvider.mm
Comment (cbracken) [inline shell/platform/darwin/macos/framework/Source/FlutterViewEngineProvider.mm:30]: See notes below on method naming/existence.

File: shell/platform/darwin/macos/framework/Source/FlutterKeyboardManager.mm
Comment (cbracken) [inline shell/platform/darwin/macos/framework/Source/FlutterKeyboardManager.mm:60]: Why weak? It might be worth documenting who's guaranteeing ownership in this relationship.

File: shell/platform/darwin/macos/framework/Source/FlutterEngine.mm
Comment (cbracken) [inline shell/platform/darwin/macos/framework/Source/FlutterEngine.mm:200]: Technically this isn't needed since `controller.viewLoaded` will return nil, `[controller loadView]` will be a no-op, and `controller.flutterView` will return nil... but I'm kind of tempted to say leave it to make the code easier to reason about anyway.

**Review by cbracken — COMMENTED** [review]
Overall idea-wise, looks good. Mostly style/etc. nitpicks. Agree with @goderbauer we should avoid adding more support/convenience for the idea of 'default' view.

File: shell/platform/darwin/macos/framework/Source/FlutterKeyboardManager.mm
Comment (dkwingsmt) [reply shell/platform/darwin/macos/framework/Source/FlutterKeyboardManager.mm:60]: This `viewDelegate` is actually the FVC, and it must be weak because the FVC owns this `FlutterKeyboardManager`, and this cyclic ownership causes memory leak.

As I've documented in [FlutterViewController](https://github.com/flutter/engine/blob/main/shell/platform/darwin/macos/framework/Headers/FlutterViewController.h#L42), FVCs strongly reference FE, and FE weakly references FVCs. The user should hold the only strong reference to FVC (in practice, that strong reference is then moved to the NSWindow). If that  strong reference is no more, then the FVC is deallocated by the system, when removes itself from the FE. Therefore, this memory leak even has behavioral significance.

File: shell/platform/darwin/macos/framework/Source/FlutterEngine.mm
Comment (dkwingsmt) [reply shell/platform/darwin/macos/framework/Source/FlutterEngine.mm:723]: It depends on which API you're using. 

If you're using `setViewController`, then the resulting FVC will always have the default ID, since `setViewController` always operates the default view.

If you're using `addViewController`, then the following FVCs will never have the default ID again. This doesn't hurt, because since you're using a new API anyway, it should be safe to assume you're not relying on the default API suite.

**Review by loic-sharma — CHANGES_REQUESTED** [review]
Please see multi-window chat. The embedder needs to notify the framework whether it can make single view assumptions or not. The current API does not let the embedder distinguish between these scenarios:

1. Current default app - The framework can make single view assumptions
2. Add-to-app with multiple views - For example, a native app where each item in a list is a Flutter view. The framework should not make single view assumptions.

I don't think the embedder can determine whether the framework can make single view assumptions with the current API.

File: shell/platform/darwin/macos/framework/Source/FlutterEngine.mm
Comment (dkwingsmt) [reply shell/platform/darwin/macos/framework/Source/FlutterEngine.mm:200]: Sorry. Are you suggesting me to keep the nil check of `controller` or not?

File: shell/platform/darwin/macos/framework/Headers/FlutterEngine.h
Comment (dkwingsmt) [reply shell/platform/darwin/macos/framework/Headers/FlutterEngine.h:111]: I'm moving adding and removing to the internal header for now. I can't assert it because there are test cases that test adding a 2nd FVC.

File: shell/platform/darwin/macos/framework/Headers/FlutterEngine.h
Comment (dkwingsmt) [reply shell/platform/darwin/macos/framework/Headers/FlutterEngine.h:106]: I think this detail is actually pretty important because it means that the first `addViewController` is equivalent to `setViewController`, which matters if people are migrating from `setViewController` to `addViewController`. To put it in another way, if you do a `addViewController` first for your engine, then `setViewController` always operates that FVC.

File: shell/platform/darwin/macos/framework/Headers/FlutterPluginRegistrarMacOS.h
Comment (dkwingsmt) [reply shell/platform/darwin/macos/framework/Headers/FlutterPluginRegistrarMacOS.h:44]: I'm postponing all deprecation to another day.

File: shell/platform/darwin/macos/framework/Source/FlutterCompositorTest.mm
Comment (dkwingsmt) [reply shell/platform/darwin/macos/framework/Source/FlutterCompositorTest.mm:14]: I'm not sure which header to put the constexpr to, because the files that rely on this value might not all import the same header. In this way I can put the declaration in any header file and just use it. Also, you can say in this way we're hiding the actual value of this magic value.

Do you think I should try to change it?

File: shell/platform/darwin/macos/framework/Source/FlutterEngine.mm
Comment (dkwingsmt) [reply shell/platform/darwin/macos/framework/Source/FlutterEngine.mm:481]: `deregisterViewControllerForId` is an internal method, and it really boils down to where to put that assertion. Doesn't matter much.

As of `viewControllerForId`, it's a public method, and I think it should really be allowed to return nil. It's just like querying a hashtable.

File: shell/platform/darwin/macos/framework/Source/FlutterEngine.mm
Comment (dkwingsmt) [reply shell/platform/darwin/macos/framework/Source/FlutterEngine.mm:703]: Great find! I've added a TODO to it.

File: shell/platform/darwin/macos/framework/Headers/FlutterEngine.h
Comment (dkwingsmt) [reply shell/platform/darwin/macos/framework/Headers/FlutterEngine.h:93]: Postponing deprecation.

File: shell/platform/darwin/macos/framework/Headers/FlutterPluginRegistrarMacOS.h
Comment (dkwingsmt) [reply shell/platform/darwin/macos/framework/Headers/FlutterPluginRegistrarMacOS.h:45]: Postponing deprecation.

**Comment by dkwingsmt** [comment]
@loic-sharma @goderbauer @cbracken Can you review again?

File: shell/platform/darwin/macos/framework/Headers/FlutterViewController.h
Comment (loic-sharma) [inline shell/platform/darwin/macos/framework/Headers/FlutterViewController.h:92]: Should there be an empty line between this comment and the previous method?

File: shell/platform/darwin/macos/framework/Source/FlutterEngine.mm
Comment (loic-sharma) [inline shell/platform/darwin/macos/framework/Source/FlutterEngine.mm:88]:
```suggestion
 * An internal method that adds the view controller with the given ID.
```

File: shell/platform/darwin/macos/framework/Source/FlutterEngine.mm
Comment (loic-sharma) [reply shell/platform/darwin/macos/framework/Source/FlutterEngine.mm:484]: Sounds good, marking as resolved

File: shell/platform/darwin/macos/framework/Source/FlutterEngine.mm
Comment (loic-sharma) [inline shell/platform/darwin/macos/framework/Source/FlutterEngine.mm:499]: As per [this comment](https://github.com/flutter/engine/pull/37976/files#r1096338601), consider:

```suggestion
  NSAssert(controller != nil && controller.id == viewId,
           @"The stored controller has unexpected view ID.");
```

File: shell/platform/darwin/macos/framework/Source/FlutterEngine.mm
Comment (loic-sharma) [inline shell/platform/darwin/macos/framework/Source/FlutterEngine.mm:521]: Should this use the `currentController`'s ID?

File: shell/platform/darwin/macos/framework/Source/FlutterEngine_Internal.h
Comment (loic-sharma) [inline shell/platform/darwin/macos/framework/Source/FlutterEngine_Internal.h:81]: Consider generalizing this comment as there are some metrics unrelated to size:

```suggestion
 * Informs the engine that the specified view controller's window metrics have changed.
```

**Review by loic-sharma — APPROVED** [review]
LGTM but please also get an approval from someone that's more familiar with the macOS embedder.

File: shell/platform/darwin/macos/framework/Source/FlutterEngine.mm
Comment (dkwingsmt) [reply shell/platform/darwin/macos/framework/Source/FlutterEngine.mm:499]: Well I thought in that comment we agreed that it's ok for this method to return null?

File: shell/platform/darwin/macos/framework/Source/FlutterEngine.mm
Comment (dkwingsmt) [reply shell/platform/darwin/macos/framework/Source/FlutterEngine.mm:521]: The `setViewController` method should only operate on the default view ID. The `currentController` is bound to have the default view ID, although yeah I can assert here.

File: shell/platform/darwin/macos/framework/Headers/FlutterEngine.h
Comment (goderbauer) [inline shell/platform/darwin/macos/framework/Headers/FlutterEngine.h:85]: nit: seems odd to talk about the "first" view here since this can only have one (for now)?

File: shell/platform/darwin/macos/framework/Source/FlutterEngine_Internal.h
Comment (goderbauer) [inline shell/platform/darwin/macos/framework/Source/FlutterEngine_Internal.h:57]: nit: remove extra space?

**Review by goderbauer — APPROVED** [review]
LGTM

File: shell/platform/darwin/macos/framework/Source/FlutterEngine.mm
Comment (loic-sharma) [reply shell/platform/darwin/macos/framework/Source/FlutterEngine.mm:490]: Right, but from my understanding this assertion fails instead of returning `null`.

File: shell/platform/darwin/macos/framework/Source/FlutterEngine.mm
Comment (loic-sharma) [reply shell/platform/darwin/macos/framework/Source/FlutterEngine.mm:514]: Good point. An assertion and maybe a quick comment seems like a good idea :)

File: shell/platform/darwin/macos/framework/Headers/FlutterPluginRegistrarMacOS.h
Comment (stuartmorgan) [inline shell/platform/darwin/macos/framework/Headers/FlutterPluginRegistrarMacOS.h:45]: I don't know if it's the same as the internal breakage, but this broke `file_selector_macos`, and would potentially break any other Swift plugin as well. In Obj-C the same syntax can be used to call methods and access properties, but that's not the case in Swift, so changing a property to a method is a breaking change, which would make supporting both `master` and `stable` with a single plugin impossible. We should not change this without a *very* compelling reason.

Comment (dkwingsmt) [reply shell/platform/darwin/macos/framework/Headers/FlutterPluginRegistrarMacOS.h:45]: Yes, this is exactly why this PR was reverted. I did this only because I thought it was equivalent (for objC). I'll change it back to proprty syntax.

**Pull request merged**
